man/man3/disk.3
.TH DISK 3
.SH NAME
opendisk, Disk \- generic disk device interface
.SH SYNOPSIS
.nf
.ft L
#include <u.h>
#include <libc.h>
#include <disk.h>
.ft
.PP
.ft L
typedef struct Disk {
	char *prefix;
	char part[NAMELEN];
	int fd, wfd, ctlfd, rdonly;
	int type;
	vlong secs, secsize, size, offset;
	int c, h, s;
} Disk;
.ft
.PP
.B
Disk* opendisk(char *file, int rdonly, int noctl)
.SH DESCRIPTION
These routines provide a simple way to gather
and use information about 
disks and disk partitions,
as well as plain files.
.PP
.I Opendisk
opens
.I file
for reading and stores the file descriptor in
the 
.B fd
field of the
.B Disk
structure.
If 
.I rdonly
is not set, 
.I opendisk
also opens
.I file
for writing and stores that file descriptor in
.BR wfd .
The two file descriptors are kept separate to
help prevent accidents.
.PP
If
.I noctl
is not set, 
.I opendisk
looks for a
.B ctl
file in the same directory as the
disk file;
if it finds one, it declares
the disk to be
an
.I sd
device,
setting the
.B type
field in the
.B Disk
structure
to
.BR Tsd .
If the passed
.I file
is named
.BI fd n disk \fR,
it looks for a file
.BI fd n ctl \fR,
and if it finds that,
declares the disk to be
a floppy disk, of type
.BR Tfloppy .
If either
control
file is found, it is opened for reading
and writing, and the resulting file descriptor
is saved as 
.BR ctlfd .
Otherwise the returned disk
has type
.BR Tfile .
.PP
.I Opendisk
then stats the file and stores its length in
.BR size .
If the disk is an
.I sd
partition, 
.I opendisk
reads the sector size from the
control
file and stores it in 
.BR secsize ;
otherwise the sector size is assumed to be 512,
as is the case for floppy disks.
.I Opendisk
then stores the disk size measured in sectors in
.BR secs .
.PP
If the disk is an
.I sd
partition, 
.I opendisk 
parses the
control
file to find the partition's offset
within its disk;
otherwise it sets
.B offset
to zero.
If the disk is an ATA disk,
.I opendisk
reads
the disk geometry (number of cylinders, heads, and sectors)
from the 
.B geometry
line in the
.I sd
control file;
otherwise it sets these to zero as well.
.B Name
is initialized with the base name of
the disk partition, and is useful for forming messages to the
.I sd
control file.
.B Prefix
is set to the passed filename without
the 
.B name
suffix.
.PP
The IBM PC BIOS interface allocates
10 bits for the number of cylinders, 8 for 
the number of heads, and 6 for the number of sectors per track.
Disk geometries are not quite so simple
anymore, but to keep the interface useful,
modern disks and BIOSes present geometries
that still fit within these constraints.
These numbers are still used when partitioning
and formatting disks.
.I Opendisk
employs a number of heuristics to discover this
supposed geometry and store it in the
.BR c ,
.BR h ,
and
.B s
fields.
Disk offsets in partition tables and
in FAT descriptors are stored in a form
dependent upon these numbers, so
.I opendisk
works hard to report numbers that
agree with those used by other operating
systems; the numbers bear little or no resemblance
to reality.
.SH SOURCE
.B \*9/src/libdisk/disk.c
.SH SEE ALSO
Plan 9's
\fIfloppy\fR(3) and \fIsd\fR(3)
.SH BUGS
Disks on Unix systems do not present the interface
that 
.I opendisk
expects, so
.I opendisk
will give them type
.BR Tfile .


CHANGES
November 26, 2005
	9term: add button 3 menu
	include: move utf.h, fmt.h from u.h to libc.h
	dist: print summary w/ errors during INSTALl

November 25, 2005
	root: add Makefile and configure for the blind,
		rename NOTES to README	

November 21, 2005
	lib9, 9pserve, vac: 9P2000.u fixes (Lucho Ionkov)

November 15, 2005
	libdraw: more X11 fixes
	libc.h: sched_yield NetBSD tweak (Anselm Garbe)

November 12, 2005
	faces: fix face location
	jpg, strings: remove references to /fd/

November 12, 2005
	various: X11 fixes

November 11, 2005
	9term: more fixes (Tim Wiess)

November 10, 2005
	9term: handle \r, other fixes (Tim Wiess)
	auxstats: new Darwin tweaks

November 9, 2005
	libthread: fix OS X context switch code

November 7, 2005
	vbackup: various fixes

November 6, 2005
	9term: new port directly from rio
	include/libc.h: p9syslog, not syslog in varargck
	libauthsrv: close fd, memory leak in authdial
	libdraw: X11 debugging helps
	lib9: add getcallerpc-arm (Ron Minnich)
	auxstats: Fix FreeBSD 6 build (William Josephson)

November 4, 2005
	stats: add -C to argchars (Lou Kamenov)
	libthread: ARM fixes (Ron Minnich), getcontext fixes

November 2, 2005
	vbackup: longer timeout on BSD mount
	libthread: Linux ARM fixes (Ron Minnich)

November 1, 2005
	various: ctype sign fixes (Tom Miller)
	libfmt: avoid overflow on %f
	libthread: Linux ARM fixes (Ron Minnich), FreeBSD 5 fixes
	libventi: various cleanup

October 31, 2005
	acme: preparation for multiline tags (Paul Lalonde)
	bc: 64-bit safe (Taj Khattra)
	faces: initial port (John Cummings)
	stats: add -F option (Lou Kamenov)
	various: ctype sign fixes (Tom Miller)

October 30, 2005
	auxstats: more OS X fixes (Jeff Sickel)	

October 29, 2005
	9p: clunk fids before exiting
	mk: handle MKSHELL type correctly
	upas, acme mail: add initial port (John Cummings)
	vbackup: silence compiler warnings
	venti: silence compiler warnings
	lib9: remove reference to ffork
		print correct file name in truerand error
	libdisk: add Disk routines
	libdiskfs: silence compiler warnings
	libframe: add noredraw flag
	libhtml: many fixes
	libsunrpc: silence compiler warnings

October 18, 2005
	9l: add autoframework (like autolib) for OS X (Jeff Sickel)
	page: use "9 file" always
	ps, psu: narrower display
	stats: better OS X support (Jeff Sickel)

October 17, 2005
	cb: undef old type macros before redefining.

October 5, 2005
	stats: better OS X support (Jeff Sickel) 

September 30, 2005
	libhtml: more &names and numbers (Erik Quanstrom)

September 29, 2005
	libc.h: add #pragmas for varargck, #ifdef'ed out.

September 26, 2005
	libthread: print thread info on SIGQUIT, SIGINFO, logging
	acme: add more thread names

September 21, 2005
	9term: fix SunOS typo (Bengt Kleberg, glaive)

September 19, 2005
	libhtml: more &names (Erik Quanstrom)

September 19, 2005
	lib9/truerand: OpenBSD lets you open /dev/random
		but not read it! (Chad Dougherty, Mechiel Lukkien)

September 15, 2005
	libdraw/x11-event: Ignore unexpected events (Rob Pike)
	jpg: do not revert to CMAP for second file (Rob Pike)

September 13, 2005
	9pserve: signedness bug

September 12, 2005
	dist: various attempts at Debian packaging
	lib/moveplan9.sh: allow specification of old directory
	Getdir: interpret tags with $ in them
	mpm: use g++ as LD, add tmac.pm to tmac directory
	include/libc.h, fcall.h: add 9P2000.u extensions
	9pserve: add 9P2000.u extensions
	vac: add 9P2000.u extensions, signedness bug

September 9, 2005
	lc: add -p like man page says
	various: insert (uchar) casts for ctype arguments
		for broken NetBSD.

September 4, 2005
	dist/main.html: note x86 OpenBSD

August 31, 2005
	file: spell IA-64 correctly (Zoltan Jarai)

August 30, 2005
	dist/debian, dist/pkg: start playing with building packages
	dist/isum: ignore 'loop not entered at top' (SunOS cc)
	dist/ss.html: add screen shots
	dist/unix.html: update VN license summary
	9l: no need to handle OpenBSD specially
	file: handle #! scripts, ELF better
	font/lucidasans/bold*unicode.7.font: more subfonts
	g: search .pl, .cgi
	hget: add hget(1)
	lp: add preliminary lp (not expected to work yet)
	src: use Plan 9 file(1)

August 12, 2005
	lib/moveplan9.files: add bin/stack
	include/libc.h: use DM constants from v9fs, add DMTMP
	libthread: add Linux-power-asm.s
	cb: add

August 11, 2005
	various: silence warnings
	unix/*: note new BSD license from Vita Nuova
	libdiskfs: add UFS2 support
	libthread: add NetBSD power assembly
	9term: send tty's current interrupt character on DEL keystroke
	rc: add limited rfork command, make sure errors propagate
	venti(8): document bloom line in config
	vftp: add get command
	vnfs: use strtoull for arguments instead of atoi.

August 10, 2005
	src/cmd/mkfile: do not build lp yet
	dist/addsrclinks, fixurls: use /usr/bin/env perl to find perl (NetBSD)
	lib9/_p9dir: avoid 32-bit overflow in old Linux size calculation
	lib9/truerand: use /dev/srandom on OpenBSD
	libthread: annoying wrappers for OpenBSD
	libmp: different assembly comments for old GNU as
	9p: set exit status
	file: add, appease gcc
	getflags: added
	mk: decide mtime-related race conservatively
	postscript/text2post: added
	rcmain: make fn cd safer for use with rc -e (in mk)
	rio: turn off debugging prints, remove 9wm.man, fixes from Andrey
	vbackup: NetBSD fixes for mount-BSD.c
	venti/srv: add <sys/stat.h> header for FreeBSD

July 28, 2005
	acme: subtle winunlock bug fix
	libdiskfs: fix double, triple indirect block reading for FFS
	vftp: new program
	vbackup: coalesce mount-*BSD files.

July 27, 2005
	include/u.h, src/libthread/thread.c: SunOS 10 attempt
	libdraw: less graceful, less raceful shutdown
	libthread: avoid C library races during threadexitsall
	awd: cannot use hostname -s on SunOS
	9pserve, srv: use correct type for NOFID

July 26, 2005
	various: fixes for SunOS from Bengt Kleberg

July 25, 2005
	vac: avoid infinite redirection while sniffing type of archive

July 24, 2005
	venti: many minor changes, updates

src/libdisk/disk.c
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <ctype.h>
#include <disk.h>

static Disk*
mkwidth(Disk *disk)
{
	char buf[40];

	sprint(buf, "%lld", disk->size);
	disk->width = strlen(buf);
	return disk;
}

/*
 * Discover the disk geometry by various sleazeful means.
 * 
 * First, if there is a partition table in sector 0,
 * see if all the partitions have the same end head
 * and sector; if so, we'll assume that that's the 
 * right count.
 * 
 * If that fails, we'll try looking at the geometry that the ATA
 * driver supplied, if any, and translate that as a
 * BIOS might. 
 * 
 * If that too fails, which should only happen on a SCSI
 * disk with no currently defined partitions, we'll try
 * various common (h, s) pairs used by BIOSes when faking
 * the geometries.
 */
typedef struct Table  Table;
typedef struct Tentry Tentry;
struct Tentry {
	uchar	active;			/* active flag */
	uchar	starth;			/* starting head */
	uchar	starts;			/* starting sector */
	uchar	startc;			/* starting cylinder */
	uchar	type;			/* partition type */
	uchar	endh;			/* ending head */
	uchar	ends;			/* ending sector */
	uchar	endc;			/* ending cylinder */
	uchar	xlba[4];			/* starting LBA from beginning of disc */
	uchar	xsize[4];		/* size in sectors */
};
enum {
	Toffset		= 446,		/* offset of partition table in sector */
	Magic0		= 0x55,
	Magic1		= 0xAA,
	NTentry		= 4,
};
struct Table {
	Tentry	entry[NTentry];
	uchar	magic[2];
};
static int
partitiongeometry(Disk *disk)
{
	char *rawname;
	int i, h, rawfd, s;
	uchar buf[512];
	Table *t;

	t = (Table*)(buf + Toffset);

	/*
	 * look for an MBR first in the /dev/sdXX/data partition, otherwise
	 * attempt to fall back on the current partition.
	 */
	rawname = malloc(strlen(disk->prefix) + 5);	/* prefix + "data" + nul */
	if(rawname == nil)
		return -1;

	strcpy(rawname, disk->prefix);
	strcat(rawname, "data");
	rawfd = open(rawname, OREAD);
	free(rawname);
	if(rawfd >= 0
	&& seek(rawfd, 0, 0) >= 0
	&& readn(rawfd, buf, 512) == 512
	&& t->magic[0] == Magic0
	&& t->magic[1] == Magic1) {
		close(rawfd);
	} else {
		if(rawfd >= 0)
			close(rawfd);
		if(seek(disk->fd, 0, 0) < 0
		|| readn(disk->fd, buf, 512) != 512
		|| t->magic[0] != Magic0
		|| t->magic[1] != Magic1) {
			return -1;
		}
	}

	h = s = -1;
	for(i=0; i<NTentry; i++) {
		if(t->entry[i].type == 0)
			continue;

		t->entry[i].ends &= 63;
		if(h == -1) {
			h = t->entry[i].endh;
			s = t->entry[i].ends;
		} else {
			/*
			 * Only accept the partition info if every
			 * partition is consistent.
			 */
			if(h != t->entry[i].endh || s != t->entry[i].ends)
				return -1;
		}
	}

	if(h == -1)
		return -1;

	disk->h = h+1;	/* heads count from 0 */
	disk->s = s;	/* sectors count from 1 */
	disk->c = disk->secs / (disk->h*disk->s);
	disk->chssrc = Gpart;
	return 0;
}

/*
 * If there is ATA geometry, use it, perhaps massaged.
 */
static int
drivergeometry(Disk *disk)
{
	int m;

	if(disk->c == 0 || disk->h == 0 || disk->s == 0)
		return -1;

	disk->chssrc = Gdisk;
	if(disk->c < 1024)
		return 0;

	switch(disk->h) {
	case 15:
		disk->h = 255;
		disk->c /= 17;
		return 0;

	default:
		for(m = 2; m*disk->h < 256; m *= 2) {
			if(disk->c/m < 1024) {
				disk->c /= m;
				disk->h *= m;
				return 0;
			}
		}

		/* set to 255, 63 and be done with it */
		disk->h = 255;
		disk->s = 63;
		disk->c = disk->secs / (disk->h * disk->s);
		return 0;
	}
	return -1;	/* not reached */
}

/*
 * There's no ATA geometry and no partitions.
 * Our guess is as good as anyone's.
 */
static struct {
	int h;
	int s;
} guess[] = {
	64, 32,
	64, 63,
	128, 63,
	255, 63,
};
static int
guessgeometry(Disk *disk)
{
	int i;
	long c;

	disk->chssrc = Gguess;
	c = 1024;
	for(i=0; i<nelem(guess); i++)
		if(c*guess[i].h*guess[i].s >= disk->secs) {
			disk->h = guess[i].h;
			disk->s = guess[i].s;
			disk->c = disk->secs / (disk->h * disk->s);
			return 0;
		}

	/* use maximum values */
	disk->h = 255;
	disk->s = 63;
	disk->c = disk->secs / (disk->h * disk->s);
	return 0;
}

static void
findgeometry(Disk *disk)
{
	if(partitiongeometry(disk) < 0
	&& drivergeometry(disk) < 0
	&& guessgeometry(disk) < 0) {	/* can't happen */
		print("we're completely confused about your disk; sorry\n");
		assert(0);
	}
}

static Disk*
openfile(Disk *disk)
{
	Dir *d;

	if((d = dirfstat(disk->fd)) == nil){
		free(disk);
		return nil;
	}

	disk->secsize = 512;
	disk->size = d->length;
	disk->secs = disk->size / disk->secsize;
	disk->offset = 0;
	free(d);

	findgeometry(disk);
	return mkwidth(disk);
}

static Disk*
opensd(Disk *disk)
{
	Biobuf b;
	char *p, *f[10];
	int nf;

	Binit(&b, disk->ctlfd, OREAD);
	while(p = Brdline(&b, '\n')) {
		p[Blinelen(&b)-1] = '\0';
		nf = tokenize(p, f, nelem(f));
		if(nf >= 3 && strcmp(f[0], "geometry") == 0) {
			disk->secsize = strtoll(f[2], 0, 0);
			if(nf >= 6) {
				disk->c = strtol(f[3], 0, 0);
				disk->h = strtol(f[4], 0, 0);
				disk->s = strtol(f[5], 0, 0);
			}
		}
		if(nf >= 4 && strcmp(f[0], "part") == 0 && strcmp(f[1], disk->part) == 0) {
			disk->offset = strtoll(f[2], 0, 0);
			disk->secs = strtoll(f[3], 0, 0) - disk->offset;
		}
	}

	
	disk->size = disk->secs * disk->secsize;
	if(disk->size <= 0) {
		strcpy(disk->part, "");
		disk->type = Tfile;
		return openfile(disk);
	}

	findgeometry(disk);
	return mkwidth(disk);
}

Disk*
opendisk(char *disk, int rdonly, int noctl)
{
	char *p, *q;
	Disk *d;

	d = malloc(sizeof(*d));
	if(d == nil)
		return nil;

	d->fd = d->wfd = d->ctlfd = -1;
	d->rdonly = rdonly;

	d->fd = open(disk, OREAD);
	if(d->fd < 0) {
		werrstr("cannot open disk file");
		free(d);
		return nil;
	}

	if(rdonly == 0) {
		d->wfd = open(disk, OWRITE);
		if(d->wfd < 0)
			d->rdonly = 1;
	}

	if(noctl)
		return openfile(d);

	p = malloc(strlen(disk) + 4);	/* 4: slop for "ctl\0" */
	if(p == nil) {
		close(d->wfd);
		close(d->fd);
		free(d);
		return nil;
	}
	strcpy(p, disk);

	/* check for floppy(3) disk */
	if(strlen(p) >= 7) {
		q = p+strlen(p)-7;
		if(q[0] == 'f' && q[1] == 'd' && isdigit((uchar)q[2]) && strcmp(q+3, "disk") == 0) {
			strcpy(q+3, "ctl");
			if((d->ctlfd = open(p, ORDWR)) >= 0) {
				*q = '\0';
				d->prefix = p;
				d->type = Tfloppy;
				return openfile(d);
			}
		}
	}

	/* attempt to find sd(3) disk or partition */
	if(q = strrchr(p, '/'))
		q++;
	else
		q = p;

	strcpy(q, "ctl");
	if((d->ctlfd = open(p, ORDWR)) >= 0) {
		*q = '\0';
		d->prefix = p;
		d->type = Tsd;
		d->part = strdup(disk+(q-p));
		if(d->part == nil){
			close(d->ctlfd);
			close(d->wfd);
			close(d->fd);
			free(p);
			free(d);
			return nil;
		}
		return opensd(d);
	}

	*q = '\0';
	d->prefix = p;
	/* assume we just have a normal file */
	d->type = Tfile;
	return openfile(d);
}

